cmd/cozystack-operator/main.go

```diff
@@ -0,0 +1,363 @@
+	return nil
+}
+
+// parsePlatformSource parses the source URL and returns the type, repository URL, and reference
+// Supports formats:
+//   - oci://registry.example.com/repo@sha256:digest
+//   - oci://registry.example.com/repo (ref will be empty)
+//   - git://github.com/user/repo@branch
+//   - git://github.com/user/repo (ref will default to "main")
+//   - https://github.com/user/repo@branch (treated as git)
+func parsePlatformSource(sourceURL string) (sourceType, repoURL, ref string, err error) {
+	// Normalize the URL by trimming whitespace
+	sourceURL = strings.TrimSpace(sourceURL)
+
+	// Check for oci:// prefix
+	if strings.HasPrefix(sourceURL, "oci://") {
+		// Remove oci:// prefix
+		rest := strings.TrimPrefix(sourceURL, "oci://")
+
+		// Check for @sha256: digest (look for @ followed by sha256:)
+		// We need to find the last @ before sha256: to handle paths with @ symbols
+		sha256Idx := strings.LastIndex(rest, "@sha256:")
+		if sha256Idx != -1 {
+			repoURL = "oci://" + rest[:sha256Idx]
+			ref = rest[sha256Idx+1:] // sha256:digest
+		} else {
+			// Check for @ without sha256: (might be a tag)
+			if atIdx := strings.LastIndex(rest, "@"); atIdx != -1 {
+				// Could be a tag, but for OCI we expect sha256: digest
+				// For now, treat everything after @ as the ref
+				repoURL = "oci://" + rest[:atIdx]
+				ref = rest[atIdx+1:]
+			} else {
+				repoURL = "oci://" + rest
+				ref = "" // No digest specified
+			}
+		}
+		return "oci", repoURL, ref, nil
+	}
+
+	// Check for git:// prefix or treat as git for http/https
+	if strings.HasPrefix(sourceURL, "git://") || strings.HasPrefix(sourceURL, "http://") || strings.HasPrefix(sourceURL, "https://") || strings.HasPrefix(sourceURL, "ssh://") {
+		// Parse URL to extract ref if present
+		parsedURL, err := url.Parse(sourceURL)
+		if err != nil {
+			return "", "", "", fmt.Errorf("invalid URL: %w", err)
+		}
+
+		// Check for @ref in the path (e.g., git://host/path@branch)
+		path := parsedURL.Path
+		if idx := strings.LastIndex(path, "@"); idx != -1 {
+			repoURL = fmt.Sprintf("%s://%s%s", parsedURL.Scheme, parsedURL.Host, path[:idx])
+			if parsedURL.RawQuery != "" {
+				repoURL += "?" + parsedURL.RawQuery
+			}
+			ref = path[idx+1:]
+		} else {
+			// Default to main branch if no ref specified
+			repoURL = sourceURL
+			ref = "main"
+		}
+
+		// Normalize git:// to https:// for GitRepository
+		if strings.HasPrefix(repoURL, "git://") {
+			repoURL = strings.Replace(repoURL, "git://", "https://", 1)
+		}
+
+		return "git", repoURL, ref, nil
+	}
+
+	return "", "", "", fmt.Errorf("unsupported source URL scheme (expected oci:// or git://): %s", sourceURL)
+}
+
+// generateOCIRepository creates an OCIRepository resource
+func generateOCIRepository(name, repoURL, digest string) (*unstructured.Unstructured, error) {
+	obj := &unstructured.Unstructured{}
+	obj.SetAPIVersion("source.toolkit.fluxcd.io/v1")
+	obj.SetKind("OCIRepository")
+	obj.SetName(name)
+	obj.SetNamespace("cozy-system")
```

Comment:
This isn't necessary. You can use a pointer to a typed object as the return value
`(*fluxv2.OCIRepository, error)`.